File: algorithms/cpp/144.cpp
#include <iostream>
#include <vector>
#include <stack>
using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution
{
public:
    vector<int> preorderTraversal(TreeNode* root)
    {
        vector<int> ans;
        stack<struct TreeNode *> s;
        struct TreeNode *p = root;
        while ( p || !s.empty() )
        {
            while ( p )
            {
                s.push(p);
                ans.push_back(p->val);
                p = p->left;
            }
            p = s.top();
            s.pop();
            p = p->right;
        }
        return ans;
    }
};

int main()
{
    return 0;
}

File: algorithms/python/144.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def preorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        rls, stack = [], []
        p = root
        while p or stack:
            while p:
                rls.append(p.val)
                stack.append(p)
                p = p.left
            p = stack.pop()
            p = p.right
        return rls
